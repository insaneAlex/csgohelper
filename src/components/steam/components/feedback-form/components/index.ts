export {FeedbackHeader} from './feedback-header';
export {FeedbackSuccess} from './feedback-success';
