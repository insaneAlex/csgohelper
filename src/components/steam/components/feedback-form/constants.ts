export const NAME_FIELD = 'name';
export const TEXT_FIELD = 'text';
