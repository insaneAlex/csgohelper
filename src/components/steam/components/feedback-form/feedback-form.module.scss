@import 'styles/index';

.wrapper {
  flex-direction: column;
  min-height: 350px;
  width: 650px;
  margin: 10px auto;

  @include box-container;

  .header {
    margin: 0 20px 25px;
    text-align: center;
    font-size: 18px;
  }

  .header,
  .input,
  .text {
    @include body-short;
  }

  .form {
    display: flex;
    flex-direction: column;
    gap: 10px;

    .input,
    .text {
      background-color: $box-element-color;
      border: none;
      padding: 10px;
      @include body-short;
    }

    .input {
      height: 40px;
    }

    .text {
      resize: none;
    }

    .error {
      text-align: left;
      color: $dark-red;
      font-size: 13px;
      font-weight: bold;
      font-family: 'Courier New', Courier, monospace;
    }
  }
}

@include large-screen {
  .wrapper {
    width: auto;
    margin: 10px;
  }
}
