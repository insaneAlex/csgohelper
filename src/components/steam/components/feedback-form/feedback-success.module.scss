@import 'styles/variables.scss';

.wrapper {
  flex-direction: column;
  text-align: center;
  justify-content: center;
  min-height: 350px;

  @include box-container;
  @include body-short;

  @media (min-width: 769px) {
    width: 650px;
    margin: 10px auto;
  }

  .title {
    font-size: 18px;
    margin-bottom: 25px;
  }

  .text {
    margin-bottom: 25px;
  }
}
