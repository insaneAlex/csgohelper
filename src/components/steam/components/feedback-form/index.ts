export {FeedbackSuccess} from './feedback-success';
export {FeedbackForm} from './feedback-form';
