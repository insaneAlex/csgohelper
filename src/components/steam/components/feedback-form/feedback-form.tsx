import {isFeedbackLoadingSelector, postFeedbackStart} from '@/src/redux/features';
import {ChangeEvent, FC, FormEvent, useState} from 'react';
import {Button, Separator} from '@/src/components/ui';
import {useDispatch, useSelector} from 'react-redux';
import {isEmpty} from '../../helpers';

import styles from './feedback-form.module.scss';

type FormValuesType = {name?: string; text?: string; errors?: {name?: string; text?: string}};

export const FeedbackForm: FC = () => {
  const [formState, setFormState] = useState<FormValuesType>({name: '', text: ''});
  const dispatch = useDispatch();
  const isLoading = useSelector(isFeedbackLoadingSelector);
  const handleInputChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const {name, value} = e.target;

    setFormState((state) => ({...state, [name]: value, errors: {...state.errors, [name]: ''}}));
  };

  const validate = (data: FormValuesType) => {
    const errors = {} as FormValuesType;
    if (isEmpty(data.name)) {
      errors.name = 'Required field';
    }
    if (isEmpty(data.text)) {
      errors.text = 'Required field';
    }
    return errors;
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    const validationErrors = validate(formState);

    if (Object.keys(validationErrors).length === 0) {
      dispatch(postFeedbackStart({text: formState.text as string, name: formState.name as string}));
    } else {
      setFormState((state) => ({...state, errors: validationErrors}));
    }
  };

  return (
    <form className={styles.form} onSubmit={handleSubmit}>
      <input
        id="name"
        type="text"
        name="name"
        value={formState.name}
        className={styles.input}
        placeholder="Your name or e-mail"
        onChange={handleInputChange}
      />
      {formState?.errors?.name && <div className={styles.error}>{formState.errors.name}</div>}
      <textarea
        rows={10}
        value={formState.text}
        id="text"
        name="text"
        className={styles.text}
        placeholder="Write feedback"
        onChange={handleInputChange}
      />
      {formState?.errors?.text && <div className={styles.error}>{formState?.errors?.text}</div>}

      <Separator smallMargin />
      <Button label="Submit" isSubmit loading={isLoading} disabled={isLoading} />
    </form>
  );
};
