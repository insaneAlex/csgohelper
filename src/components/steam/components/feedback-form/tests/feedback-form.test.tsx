import {fireEvent, render, screen} from '@testing-library/react';
import {FeedbackForm} from '../feedback-form';

const routerPushMock = jest.fn();
const dispatchMock = jest.fn();

jest.mock('../../../../../services', () => ({noop: jest.fn()}));
jest.mock('next/router', () => ({useRouter: () => ({push: routerPushMock})}));
jest.mock('react-redux', () => ({useDispatch: () => dispatchMock, useSelector: jest.fn()}));

describe('FeedbackForm', () => {
  it('should render correctly', () => {
    const {container} = render(<FeedbackForm />);
    expect(container).toMatchSnapshot();
  });
  describe('when changing form value', () => {
    it('updates form state', () => {
      render(<FeedbackForm />);
      const changedValue = 'MyName';
      const nameInput = screen.getByPlaceholderText('Your name or e-mail');
      fireEvent.change(nameInput, {target: {value: changedValue}});
      expect(screen.getByTestId('feedback-form')).toHaveFormValues({name: changedValue});
    });
  });
  describe('when required fields not filled', () => {
    it('should render validation error messages', () => {
      render(<FeedbackForm />);
      const submitButton = screen.getByText('Submit');
      fireEvent.click(submitButton);
      expect(screen.getAllByText('Required field')).toHaveLength(2);
    });
  });
  describe('when entered data is correct and user submit form', () => {
    it('should call dispatch action', () => {
      render(<FeedbackForm />);
      const nameInput = screen.getByPlaceholderText('Your name or e-mail');
      const textInput = screen.getByPlaceholderText('Write feedback');
      fireEvent.change(nameInput, {target: {value: 'John'}});
      fireEvent.change(textInput, {target: {value: 'Text'}});
      fireEvent.click(screen.getByText('Submit'));
      expect(dispatchMock).toHaveBeenCalled();
    });
  });
});
