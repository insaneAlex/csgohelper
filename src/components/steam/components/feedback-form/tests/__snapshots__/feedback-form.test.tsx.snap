// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm should render correctly 1`] = `
<div>
  <h1
    class="header"
  >
    Have any thoughts or ideas? Share them with us below
  </h1>
  <form
    class="form"
    data-testid="feedback-form"
  >
    <input
      class="input"
      id="name"
      name="name"
      placeholder="Your name or e-mail"
      value=""
    />
    <textarea
      class="text"
      id="text"
      name="text"
      placeholder="Write feedback"
      rows="10"
    />
    <hr
      class="separator smallMargin"
    />
    <button
      class="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
